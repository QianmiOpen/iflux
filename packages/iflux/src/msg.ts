import mitt from 'mitt';
import { useEffect } from 'react';

export type TMsgListeners = { [name: string]: mitt.Handler };

export const msg = new mitt();

export function useMsg(listener: TMsgListeners = {}) {
  //componentDidMount
  useEffect(() => {
    const keys = Object.keys(listener);

    for (let key of keys) {
      msg.on(key, listener[key]);
    }

    //componentWillUnmount
    return () => {
      for (let key of keys) {
        msg.off(key, listener[key]);
      }
    };
  }, []);

  return msg;
}
