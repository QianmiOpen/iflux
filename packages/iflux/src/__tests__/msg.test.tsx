import React, { useState } from 'react';
import renderer, { act } from 'react-test-renderer';
import { msg, useMsg } from '../msg';

it('test msg', () => {
  msg.on('hello', hello => {
    expect(hello).toEqual('hello');
  });

  msg.emit('hello', 'hello');
});

it('test useMsg', () => {
  function Test() {
    const [text, setState] = useState('test');
    useMsg({
      hello: hello => {
        act(() => setState(hello));
      }
    });

    return <div>{text}</div>;
  }

  const tree = renderer.create(<Test />);
  expect(tree).toMatchSnapshot();

  msg.emit('hello', 'hello');
  expect(tree).toMatchSnapshot();
});
