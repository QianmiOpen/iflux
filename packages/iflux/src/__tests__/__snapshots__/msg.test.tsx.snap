// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test useMsg 1`] = `
<div>
  test
</div>
`;

exports[`test useMsg 2`] = `
<div>
  hello
</div>
`;
